File: collatz/collatz.cpp
// https://open.kattis.com/problems/collatz

#include <iostream>
#include <vector>

using namespace std;

int next(int current) {
	if (current % 2 == 0) {
		return current / 2;
	} else {
		return 3 * current + 1;
	}
}

vector<int> get_sequence(int start_number) {
	vector<int> seq;
	seq.push_back(start_number);

	int x = start_number;
	while (x != 1) {
		x = next(x);
		seq.push_back(x);
	}

	return seq;
}

int main(void) {
	while (true) {
		int A, B;
		cin >> A >> B;
		if (A == 0 && B == 0) {
			break;
		}

		vector<int> seq_A = get_sequence(A), seq_B = get_sequence(B);

		vector<int>::iterator ia, ib, ia_prev, ib_prev;
		for (
			ia = seq_A.end(), ib = seq_B.end(), ia_prev = ia - 1, ib_prev = ib - 1;
			ia > seq_A.begin() && ib > seq_B.begin() && *ia_prev == *ib_prev;
			ia = ia_prev--, ib = ib_prev--
		);

		cout << A << " needs " << (ia - seq_A.begin()) << " steps, " << B << " needs " << (ib - seq_B.begin()) << " steps, they meet at " << *ia << endl;
	}

	return 0;
}

File: collatz/collatz.go
// https://open.kattis.com/problems/collatz

package main

import "fmt"

func main() {
	var A, B, a, b uint64
	var steps_a, steps_b uint32
	var steps map[uint64]uint32
	for true {
		fmt.Scanf("%d", &A)
		fmt.Scanf("%d", &B)
		if ( (A == 0) && (B == 0) ) {
			break
		}

		steps = make(map[uint64]uint32)

		a = A
		steps_a = 0
		steps[a] = steps_a
		for (a != 1) {
			if ( (a & 1) > 0) {
				a = a + a + a + 1
			} else {
				a = a >> 1
			}
			steps_a++
			steps[a] = steps_a
		}

		b = B
		steps_b = 0
		for true {
			_, ok := steps[b]
			if (ok) {
				steps_a = steps[b]
				break
			}

			if ( (b & 1) > 0) {
				b = b + b + b + 1
			} else {
				b = b >> 1
			}

			steps_b++
		}

		fmt.Printf("%d needs %d steps, %d needs %d steps, they meet at %d\n", A, steps_a, B, steps_b, b)
	}
}